sys/net/gnrc/network_layer/ndp2/gnrc_ndp2.c

```diff
@@ -0,0 +1,567 @@
+{
+    ndp_opt_t *opt;
+    gnrc_pktsnip_t *pkt = gnrc_pktbuf_add(next, NULL, _ceil8(size),
+                                          GNRC_NETTYPE_UNDEF);
+
+    if (pkt == NULL) {
+        DEBUG("ndp2: no space left in packet buffer\n");
+        return NULL;
+    }
+    opt = pkt->data;
+    opt->type = type;
+    opt->len = (uint8_t)(pkt->size / 8);
+    return pkt;
+}
+
+static inline gnrc_pktsnip_t *_opt_l2a_build(const uint8_t *l2addr,
+                                             uint8_t l2addr_len,
+                                             gnrc_pktsnip_t *next,
+                                             uint8_t type)
+{
+    gnrc_pktsnip_t *pkt = gnrc_ndp2_opt_build(type,
+                                              sizeof(ndp_opt_t) + l2addr_len,
+                                              next);
+
+    if (pkt != NULL) {
+        ndp_opt_t *l2a_opt = pkt->data;
+
+        memset(l2a_opt + 1, 0, pkt->size - sizeof(ndp_opt_t));
+        memcpy(l2a_opt + 1, l2addr, l2addr_len);
+    }
+    return pkt;
+}
+
+gnrc_pktsnip_t *gnrc_ndp2_opt_sl2a_build(const uint8_t *l2addr,
+                                         uint8_t l2addr_len,
+                                         gnrc_pktsnip_t *next)
+{
+    assert((l2addr != NULL) && (l2addr_len != 0));
+    DEBUG("ndp2: building source link-layer address option (l2addr: %s)\n",
+          gnrc_netif_addr_to_str(addr_str, sizeof(addr_str), l2addr,
+                                 l2addr_len));
+    return _opt_l2a_build(l2addr, l2addr_len, next, NDP_OPT_SL2A);
+}
+
+gnrc_pktsnip_t *gnrc_ndp2_opt_tl2a_build(const uint8_t *l2addr,
+                                         uint8_t l2addr_len,
+                                         gnrc_pktsnip_t *next)
+{
+    assert((l2addr != NULL) && (l2addr_len != 0));
+    DEBUG("ndp2: building target link-layer address option (l2addr: %s)\n",
+          gnrc_netif_addr_to_str(addr_str, sizeof(addr_str), l2addr,
+                                 l2addr_len));
+    return _opt_l2a_build(l2addr, l2addr_len, next, NDP_OPT_TL2A);
+}
+
+gnrc_pktsnip_t *gnrc_ndp2_opt_pi_build(const ipv6_addr_t *prefix,
+                                      uint8_t prefix_len,
+                                      uint32_t valid_ltime, uint32_t pref_ltime,
+                                      uint8_t flags, gnrc_pktsnip_t *next)
+{
+    assert(prefix != NULL);
+    assert(!ipv6_addr_is_link_local(prefix) && !ipv6_addr_is_multicast(prefix));
+    assert(prefix_len <= 128);
+    gnrc_pktsnip_t *pkt = gnrc_ndp2_opt_build(NDP_OPT_PI, sizeof(ndp_opt_pi_t),
+                                              next);
+
+    if (pkt != NULL) {
+        ndp_opt_pi_t *pi_opt = pkt->data;
+
+        pi_opt->prefix_len = prefix_len;
+        pi_opt->flags = (flags & NDP_OPT_PI_FLAGS_MASK);
+        pi_opt->valid_ltime = byteorder_htonl(valid_ltime);
+        pi_opt->pref_ltime = byteorder_htonl(pref_ltime);
+        pi_opt->resv.u32 = 0;
```

Comment:
The packets in the packet buffer are never overridden (except if you actually provide data), so no.